### PR TITLE
fix[codegen]: recursive dynarray oob check

### PR DESCRIPTION
**Pull request by charles-cooper**
### What I did
fix another bug in dynarray recursion

reported by @trocher 

### Commit message

```
this commit fixes more edge cases in `abi_decode` dynarray
validation. these are bugs which were missed (or regressions) in
1f6b9433fbd524, which itself was a continuation of eb011367cc769.

there are multiple fixes contained in this commit.

- similar conceptual error as in 1f6b9433fbd524. when the
length word is out-of-bounds and its runtime is value is zero,
`make_setter` does not enter recursion and therefore there is
no oob check. an example payload which demonstrates this is in
`test_nested_invalid_dynarray_head()`. the fix is to check the
size of the static section ("embedded static size") before entering
the recursion, rather than child_type.static_size (which could be
zero). essentially, this checks that the end of the static section is
in bounds, rather than the beginning.

- the fallback case in `complex_make_setter` could be referring to a
tuple of dynamic types, which makes the tuple itself dynamic, so there
needs to be an oob check there as well.

- `static_size()` is more appropriate than `min_size()` for abi payload
validation, because you can have "valid" ABI payloads where the runtime
length of the dynamic section is zero, because the heads in the static
section all point back into the static section. this commit replaces
the `static_size()` check with `min_size()` check, everywhere.

- remove `returndatasize` check in external calls, because it gets
checked anyways during `make_setter` oob checks.

- add a comment clarifying that payloads larger than `size_bound()` get
rejected by `abi_decode` but not calldata decoding.

tests for each case, contributed by @trocher

---------

Co-authored-by: trocher <trooocher@proton.me>
```

### Cute Animal Picture

image

### CODE REVIEW

File: tests/functional/builtins/codegen/test_abi_decode.py [hunk 1]
```diff
@@ -1323,3 +1323,92 @@ def run(x: Bytes[2 * 32 + 3 * 32  + 3 * 32 * 4]):
...
+
+
+def test_static_outer_type_invalid_heads(get_contract, tx_failed):
+    code = """
+@nonpayable
+@external
+def foo(x:Bytes[320]):
+    if True:
+        a: Bytes[320-32] = b''
+        b:uint256 = 32
```
Comment (trocher) [inline tests/functional/builtins/codegen/test_abi_decode.py:1381]:
```suggestion
```

File: tests/functional/builtins/codegen/test_abi_decode.py [hunk 2]
```diff
@@ -1323,3 +1323,92 @@ def run(x: Bytes[2 * 32 + 3 * 32  + 3 * 32 * 4]):
...
+    y:DynArray[uint256, 2][2] = _abi_decode(x_mem,DynArray[uint256, 2][2])
+
+@nonpayable
+@external
+def bar(x:Bytes[320]):
+    if True:
+        a: Bytes[160] = b''
+        # write stuff here to make the call revert in case decode do
+        # an out of bound access:
+        b: uint256 = 32
```
Comment (trocher) [inline tests/functional/builtins/codegen/test_abi_decode.py:1392]:
```suggestion
        fake_head: uint256 = 32
```

File: tests/functional/codegen/calling_convention/test_external_contract_calls.py [hunk 3]
```diff
@@ -2520,12 +2520,12 @@ def foo(a: DynArray[{typ}, 3], b: String[5]):
...
 
     # Dynamic size is short by 1 byte
-    malformed = data[:264]
+    malformed = data[:136]
     with tx_failed():
         env.message_call(c1.address, data=malformed)
 
-    # Dynamic size is at least minimum (132 bytes * 2 + 2 (for 0x) = 266)
-    valid = data[:266]
+    # Dynamic size is at least minimum (66 bytes * 2 + 2 (for 0x) = 266)
```
Comment (trocher) [inline tests/functional/codegen/calling_convention/test_external_contract_calls.py:2527]:
```suggestion
    # Dynamic size is at least minimum ((4 + 64) bytes * 2 + 2 (for 0x) = 138)
```

File: tests/functional/codegen/calling_convention/test_external_contract_calls.py [hunk 4]
```diff
@@ -2520,12 +2520,12 @@ def foo(a: DynArray[{typ}, 3], b: String[5]):
     data = f"0x{sig}{encoded}"
 
     # Dynamic size is short by 1 byte
```
Comment (trocher) [inline tests/functional/codegen/calling_convention/test_external_contract_calls.py:2522]:
```suggestion
    # Static size is short by 1 byte
```

File: tests/functional/builtins/codegen/test_abi_decode.py [hunk 5]
```diff
@@ -1323,3 +1323,89 @@ def run(x: Bytes[2 * 32 + 3 * 32  + 3 * 32 * 4]):
...
+        0x00,  # 0x60
+        0x00,  # 0x80
+        0x00,  # 0xA0
+        0x00,  # 0xC0
+        0x02,  # len of outer
+    )
+    inner = (
+        0x0,  # head
+        # 0x0,  # head2
+    )
```
Comment (trocher) [inline tests/functional/builtins/codegen/test_abi_decode.py:1365]:
```suggestion
        0x20,  # head of the dynarray
        0x02,  # len of outer
    )
    inner = (
        0x0,  # head1
        # 0x0,  # head2
    )
```
reduce tests given that `min_size` is no longer used

File: tests/functional/builtins/codegen/test_abi_decode.py [hunk 6]
```diff
@@ -1323,3 +1323,89 @@ def run(x: Bytes[2 * 32 + 3 * 32  + 3 * 32 * 4]):
...
+    encoded = (
+        0x80,  # head of the static array
+        0x00,  # garbage to pass the ABI min_size check
+        0x00,  # garbage to pass the ABI min_size check
+        0x00,  # garbage to pass the ABI min_size check
+    )
+    inner = (
+        0x00,  # head of the first dynarray
+        # 0x00,  # head of the second dynarray
+    )
```
Comment (trocher) [inline tests/functional/builtins/codegen/test_abi_decode.py:1404]:
```suggestion
    encoded = (
        0x20,  # head of the static array
    )
    inner = (
        0x00,  # head of the first dynarray
        # 0x00,  # head of the second dynarray
    )
```
reduce tests given that `min_size` is no longer used

File: vyper/codegen/core.py [hunk 7]
```diff
@@ -982,7 +979,14 @@ def make_setter(left, right, hi=None):
...
 
-    return _complex_make_setter(left, right, hi=hi)
+    ret = ["seq"]
+    if hi is not None:
+        item_end = add_ofst(right, right.typ.abi_type.static_size())
+        len_check = ["assert", ["le", item_end, hi]]
+        ret.append(len_check)
+
+    ret.append(_complex_make_setter(left, right, hi=hi))
+    return IRnode.from_list(ret)
```
Comment (trocher) [inline vyper/codegen/core.py:989]: `make_setter`'s caller always takes care of it from what I saw, but it wouldn't hurt to `cache_when_complex` `right` as it is done in the other cases.

Comment (charles-cooper) [reply vyper/codegen/core.py]: https://github.com/vyperlang/vyper/pull/4091/commits/a10662ba3b1d61fc257e3bf917842587eb7e4e42

**Comment by trocher** [comment]
lgtm :)

**Review by trocher — COMMENTED** [review]
lgtm :)

**Review by cyberthirst — APPROVED** [review]

**Pull request merged**